### ISSUE TITLE
Type not narrowed  on `isinstance(obj, Callable)` check

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#2274]
When checking if an object is a `Callable` type, pyright does not narrow the type.

Sample code:
```py
from typing import Callable, Union


def some_func(obj: Union[Callable, int]):
    if isinstance(obj, Callable):
        reveal_type(obj) # Type of "obj" is "(*args: Unknown, **kwargs: Unknown) -> Unknown | int"
    else:
        reveal_type(obj) # Type of "obj" is "(*args: Unknown, **kwargs: Unknown) -> Unknown | int"
```

The above code is valid python code (disregarding the `reveal_type`), and will correctly branch based on whether or not `obj` implements `__call__`, however the type is not narrowed.

pyright correctly narrows the type if checking with `callable(obj)`, ie.

```py
from typing import Callable, Union


def some_func(obj: Union[Callable, int]):
    if callable(obj):
        reveal_type(obj) # Type of "obj" is "(*args: Unknown, **kwargs: Unknown) -> Unknown"
    else:
        reveal_type(obj) # Type of "obj" is "int"
````

Tested on `pyright 1.1.166`

**Comment by erictraut** [comment]
I didn't realize that `typing.Callable` could be used with `isinstance`. Do you happen to know if that's documented somewhere? Does it work in the same way as the `callable` function?

**Comment by erictraut** [comment]
I'll note that mypy doesn't apply type narrowing in the top sample either, but it does in the bottom. I'm guessing that the bottom pattern is the recommended (or at least the popular) approach.

**Comment by kennipj** [comment]
Not sure how `callable` works, but I do know that the isinstance check works on methods, functions, lambdas and objects with a `__call__` attribute.

To me it's intuitive that the `isinstance(obj, Callable)` check would work, as this similar to how you would check other objects for their type.

I went for this approach immediately when I wanted to check if the object was a function, but was surprised when no type narrowing was applied, but observed that the check worked just fine.

**Comment by kennipj** [comment]
Upon further experimentation, it seems both `callable(obj)` and `isinstance(obj, Callable)` check for `__call__` attribute. But both don't actually check if `__call__` is a function.

for example this here will result in a raised exception
```py
from typing import Callable, Union


class MyObj:
    __call__ = "hello"


def some_func(obj: Union[Callable, int]):
    if callable(obj):
        obj()
        # TypeError: 'str' object is not callable
    else:
        print("hi")


some_func(MyObj())
```

But pyright will complain that `MyObj()` is not actually "callable", so all good there.

**Comment by erictraut** [comment]
This will be implemented in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.167, which I just published. It will also be included in the next release of pylance.
